### ISSUE TITLE
Prepare Istio 1.18

### ISSUE THREAD

**Issue by pmerrison** [istio/istio#44075]
Modified version of #42845 (from 1.17)

These instructions were extensively re-rewritten for clarity as part of the 1.18 release process.  Reviews / comments / updates much appreciated!

---
## Introduction
This issue tracks all steps needed to prepare the release 1.18 branches/builds.

This issue was created early to help document the steps using some new automation as well as any ordering changes, etc. Please update/add additional steps taken here so we can document for future releases.

**Note that the actual coding steps done by the automation steps are left in this issue for reference only. You only need to run the automated steps, and not the individual instructions. Also note that the reference steps may be INCOMPLETE as the automation is updated, not necessarily the reference steps.**

## The Automation
The release-builder automation can be found in https://github.com/istio/release-builder, master branch.  It's best if you fork this repo into your own org, as you will need to submit PRs into the repo to trigger the automation to run on Github rather than locally.

Whilst you're there it's also worth forking all* of the other Istio repos into your own organisation, as that way you can dry-run against it.  *by all, I mean the ones specified in release/branch.sh

### Important things to know about the automation

First, spend a little time understanding the release-builder [readme](https://github.com/istio/release-builder/blob/master/README.md).  Don't worry if you don't understand all of it, it's not really necessary to in order to work through the rest of these steps, but having some understanding will help.  Here are the important things you must know:

The release directory contains three config files:
   * trigger-branch - this contains the version and step for the automation to work on
   * trigger-build - this contains the tag for the release-builder to use for whatever it's building
   * trigger-publish - this contains the tag that the release-builder will publish.

Before you start testing the automation with dry runs, update those files to whatever you need them to be.

Each time you start the build container with `make shell` you need to:
* Run `gh auth login` (watch for errors, you may need to run this command with `sudo`
You also need to remember that for not-dry-runs you need to ensure that the GITHUB_TOKEN environment variable is set **within** the build container.  The automation does not care at all if you've logged in with the `gh` CLI.  The token that you use has to have `repo` permissions.  If you don't do this, the automation will fail when trying to interact with github by pushing, creating PRs etc.

### How to run it

Update the relevant files to specify what you want, then:

```
make shell
gh auth login
GITHUB_TOKEN=my_token
REPO_ORG=myorg STEP=x ./release/branch.sh
```

This generates the changes for you to review (look at `git status` or `git diff` to see them).  When you are satisfied with what you see, run

```
make shell
REPO_ORG=myorg STEP=x DRY_RUN=false ./release/branch.sh 
```

This will cause the actual step to happen and PRs to be raised in whichever github org you specified in `REPO_ORG`

Note that the automated steps 3 and on require the new release's branch in the various repos which are created by automated step 2.  

Also note that if you're on a Mac you need to:
* Run `make shell` with `sudo`
* Run`brew install docker-credential-helper` first to avoid difficult-to-spot errors in the scripts
* Edit ~/.docker/config.json, change the line credsstore (two s) to credstore.
On Linux, I had no luck getting it to work on ubuntu (git clone would hang), but it worked fine on debian11.  That could just have been that my system had gremlins though.

### When to start: 2 weeks before branch cut date
- [x] Set up release manager team - Ensure the new team has write access to the repositories. (see [teams.yaml](https://github.com/istio/community/blob/master/org/teams.yaml)) 
- [x] Create a slack channel for the [1.18 release](https://istio.slack.com/archives/C04R6EZANU9). Edit the description/topic of the channel to include RM's for the release, expected release date, etc.
- [x] Add new label: cherrypick/release-1.18: [1.18 PR](https://github.com/istio/bots/pull/665)
- [x] Create a 1.19 milestone: [1.18 PR](https://github.com/istio/bots/pull/665)
- [ ] Send an announcement on Discuss pre-announcing branch cut (adding, but are branch cut notifications needed and where?)
- [ ] Send an announcement on [Discuss](https://discuss.istio.io/t/1-17-release-branch-cut/14764) announcing the start of branch cut and to stop merging of PRs (adding, but are branch cut notifications needed and where?)

## Automation Step 1 - Update Dependencies

This step updates dependencies in master prior to the release branch being cut

### Command to run
```
make shell
REPO_ORG=myorg STEP=1 ./release/branch.sh
```

### Legacy Manual Steps
- (**Automation step=1**) Update dependencies. This must be done **before branching** so that PRs can be easily backported later. [PR](https://github.com/istio/istio/pull/44436)
  - Run `go get github.com/envoyproxy/go-control-plane@main` to update (Added in 1.17 in https://github.com/istio/release-builder/pull/1332).
  - Run the following command in istio/istio `UPDATE_BRANCH=master ./bin/update_deps.sh; make gen` ([example](https://github.com/istio/istio/pull/36812))
  - [x] Merge PR in istio/istio. Wait for it to complete. 
  - [x] _Need to verify git code to allow commits and pulls to work for upcoming 1.18 branch cut._

### When to start: 1 week before branch cut date

This step creates release-1.18 branches in the necessary repos.  

### Command to run
```
make shell
REPO_ORG=myorg STEP=2 ./release/branch.sh
```

### Legacy Manual Steps
- (**Automation step=2**)  (new step because we are waiting on prior change to master before branching) Create a `release-1.18` branch in every Istio repo - run commands `export org=xyz ; export repo=xyz ; (git clone git@github.com:${org}/${repo}.git && cd $repo && git checkout -b release-1.18 && git push --set-upstream origin release-1.18)`.
  - istio/istio 
  - istio/pkg
  - istio/api 
  - istio/envoy  (~~**NOT done as part of release-builder**~~) ([ ] Should update code to include this again? It seems Step=4 needs it, so will add it back) [PR to add it back](https://github.com/istio/release-builder/pull/1334)
  - istio/proxy
  - istio/client-go 
  - istio/tools
  - istio/common-files 
  - istio/release-builder 
  - istio/enhancements
  - Explicitly skipped: istio-releases/pipeline, istio/tests, istio/istio.io, istio/test-infra, istio/gogo-genproto (No longer used and not done as part of release-builder [1.15 PR](https://github.com/istio/release-builder/pull/1003)) 
  - No PRs to merge as the new branches were created by automation

### Post-automation steps
- [x] Set up branch protection. Modify prow/config.yaml in test-infra to add release managers as owners for a branch. [1.18 PR](https://github.com/istio/test-infra/pull/4674)

## Automation Step 3 - Set Up Prow On Release Branches

21/4/2023 - automation for this step is currently broken with an early segmentation fault, I'll work through debugging this later

Only a Googler who has the ability to tag new images in the istio-testing gcr repo can run this step.  That means that you have three choices available:
* Find one who is willing to work through running the automation
* Puzzle out how the release-builder automation works well enough to edit out the three tagging commands below, comment them out.  Then find a googler to create the three tags and then run your edited automation
* Find a googler to create the three tags below, then work through the remaining steps below manually

### Command to run
```
make shell
REPO_ORG=myorg STEP=3 ./release/branch.sh
```

### "Legacy" Manual Steps
**Note:** Here is the 1.18 PR for the manual steps - https://github.com/istio/test-infra/pull/4675
Since the automation currently doesn't work, in order to do this manually you need to:

- Have a someone with creds for the istio-testing GCR repo create the new tags for `build-tools`, `build-tools-proxy` and `build-tools-centos`images (e.g. John Howard).  For instance (replace tags with latest):
      - `gcloud container images add-tag gcr.io/istio-testing/build-tools:master-9bab62ff1c9c252ed3189a6cb4eb0bec674ad3a1 gcr.io/istio-testing/build-tools:release-1.18-9bab62ff1c9c252ed3189a6cb4eb0bec674ad3a1`
      - `gcloud container images add-tag gcr.io/istio-testing/build-tools-proxy:master-master-9bab62ff1c9c252ed3189a6cb4eb0bec674ad3a1 gcr.io/istio-testing/build-tools-proxy:release-1.18-9bab62ff1c9c252ed3189a6cb4eb0bec674ad3a1`
      - `gcloud container images add-tag gcr.io/istio-testing/build-tools-centos:master-master-9bab62ff1c9c252ed3189a6cb4eb0bec674ad3a1 gcr.io/istio-testing/build-tools-centos:release-1.18-9bab62ff1c9c252ed3189a6cb4eb0bec674ad3a1`

Next, checkout your fork of the test-infra project (making sure it's up to date with master first), `gh auth login`, create a new branch etc.  Then:

   - `sudo make shell`
   - `go get github.com/kballard/go-shellquote` then `go get istio.io/test-infra/tools/prowgen/pkg` -- don't commit the modified go.mod and go.sum from these commands
   - Generate the jobs: `go run tools/prowgen/cmd/prowgen/main.go branch 1.18`
   - Generate the private jobs: `go run ./tools/generate-transform-jobs/main.go branch 1.18`
   - In the job configs prow/**/*-1.18.yaml, replace any image build-tools(-centos|-proxy):master with build-tools(-centos|-proxy):release-1.18.  This is easy to do in VSCode, like this:
      - <img width="304" alt="Screenshot 2023-04-21 at 10 20 54" src="https://user-images.githubusercontent.com/10532667/233598874-67476641-7129-4b72-a7ff-2fc0efa89769.png">

   - Remove duplicated `env` entries of the following in the configs (you can use a similar search in VSCode as above, but beware, it'll only match the indentation level that you specify if you have strict matching enabled.):
      ```
      - name: BUILD_WITH_CONTAINER
        value: "0"
      ```

    - Manually update [testgrid/config.yaml](https://github.com/istio/test-infra/blob/master/testgrid/config.yaml) by adding in the relevant entries for 1.18.  It's pretty obvious once you see the file what you have to do.  TBD - decide what to do with the previous entries - do we delete them?
    - Remove? - Manually update BASE_VERSION in the istio-{release} yaml from `master` to {release} 
    - Run `make gen`
    - [ ] _Update test-infra code to handle duplicate entries as well as master image changes done manually above for upcoming 1.19 branch cut._
  - [x] Merge PR in istio/test-infra. Wait for it to complete. 

## Automation Step 4

This step has a two steps:

* Run the automation
* Merge the PRs created by it, EXCEPT FOR ONE! As part of the STEP 4 PR post-submits, a new PR will be created in the istio/common-files repo with a title like: `Automator: update build-tools image@release-1.18 in istio/common-files@release-1.18`. The work in this PR is only a portion of Step 5, and merging it will cause issues with reverting the repos to the main branch common-files. **CLOSE this PR**.
 
These are the PRs created by the automation for 1.18:
* https://github.com/istio/release-builder/pull/1467
*https://github.com/istio/api/pull/2766
* https://github.com/istio/pkg/pull/790
* https://github.com/istio/tools/pull/2446
* https://github.com/istio/envoy/pull/340
* https://github.com/istio/enhancements/pull/149
* https://github.com/istio/istio/pull/44553
* https://github.com/istio/proxy/pull/4607
* https://github.com/istio/ztunnel/pull/506
* https://github.com/istio/client-go/pull/1514

### Command To Run
```
make shell
REPO_ORG=istio STEP=4 ./release/branch.sh
```

### Legacy Manual Instructions
- (**Automation step=4**) (new step since we want the automation from prior step to actually create the image after PR merges) PRS: [1 - istio/pkg](https://github.com/istio/pkg/pull/722) [2 - istio/client-go](https://github.com/istio/client-go/pull/1351) [3 - istio/tools](https://github.com/istio/tools/pull/2320) [4 - istio/common-files](https://github.com/istio/common-files/pull/754) [5 - istio/enhancements](https://github.com/istio/enhancements/pull/131) [6 - istio/release-builder](https://github.com/istio/release-builder/pull/1335) [7 - istio/istio](https://github.com/istio/istio/pull/42868) [8 - istio/api](https://github.com/istio/api/pull/2639) [9 - istio/proxy](https://github.com/istio/proxy/pull/4351) [10 - istio/envoy](https://github.com/istio/envoy/pull/338)
  - Updates istio/tools to build new release-1.18 build image (update BRANCH in docker/build-tools/build-and-push.sh. PR postsubmit will create new container images whose name will be used in the next step) 
  - Update common-files in _new release_. You first have to manually update the _common/Makefile.common.mk_ *update-common* target to point to the _new release_  (but not in common-files). In Step=5, the prow automation will actually call `make update-common` in these repos to do the actual `make update-common` against the new common-files release branch
    - istio/common-files - skipped. Covered in next step
    - istio/istio 
    - istio/pkg 
    - istio/api
    - istio/client-go
    - istio/proxy 
    - istio/release-builder 
    - istio/tools 
  - Update `CODEOWNERS` to contain only the release managers for this release. Command: `export org=xyz ; export repo=xyz ; (git clone git@github.com:${org}/${repo}.git && cd $repo && git checkout release-1.18 && git checkout -b release-1.18-codeowners && echo '* @istio/release-managers-1.18' > CODEOWNERS && git add CODEOWNERS && git commit -m 'Set release managers as CODEOWNERS for release-1.18' && git push --set-upstream origin release-1.18-codeowners)`.
    - istio/common-files  - skipped. Covered in next step
    - istio/istio 
    - istio/pkg 
    - istio/api 
    - istio/proxy
    - istio/client-go 
    - istio/tools
    - istio/release-builder
    -  istio/enhancements
  - Stop publishing `latest` tags
  - Update istio/release-builder branch changes from `master` to new release in build.sh, publish.sh and manifests. 
  
  [x] Merge PRs from STEP 4. Wait for the new build images to be created at https://gcr.io/istio-testing/build-tools.
**WARNING -- DO NOT MERGE** As part of the STEP 4 PR post-submits, a new PR will be created in the istio/common-files repo with a title like: `Automator: update build-tools image@release-1.18 in istio/common-files@release-1.18`. The work in this PR is only a portion of Step 5, and merging it will cause issues with reverting the repos to the main branch common-files. **CLOSE this PR**.

### Command to run

```
make shell
REPO_ORG=istio STEP=5 ./release/branch.sh
```

### Legacy Manual Steps
- (**Automation step=5**) (new step since we need image from prior step)
  - Update istio/common-files to set release-1.18 build image (Update the UPDATE_BRANCH in files/common/Makefile.common.mk to be the new release name
  - Update IMAGE_VERSION in files/common/scripts/setup_env.sh  to be the new build image from prior step (found at https://gcr.io/istio-testing/build-tools).
  - Also Update `CODEOWNERS` to contain only the release managers for this release. 

  [x] Merge [PR](https://github.com/istio/common-files/pull/821) from STEP 5. Wait until step 5 PR merges and all the postsubmit generated PRs (to update common files in repos) merge.

## Remaining Manual Steps

After the automation is complete, there are some manual steps that need to be completed to finish the branch cut

- [x] Fix up proxy automated job to pull from correct envoyproxy release branch if possible (set UPDATE_BRANCH). [example](https://github.com/istio/test-infra/pull/4681) 

- [x] Verify that the github.com/envoyproxy/go-control-plane matches the latest commit before github.com/envoyproxy/envoy was branched. Easiest way to determine this is to go to the new envoy branch: https://github.com/envoyproxy/envoy/tree/release/v1.26 in this case, and click on commits behind main to find the commit where the branch happened. Then go to the main branch and find that commit. Finally go to https://github.com/envoyproxy/go-control-plane/commits/main and find the last commit made before the branch commit. Update in istio and proxy repos. [1.18 istio PR(https://github.com/istio/istio/pull/44665) [1.18 proxy PR](https://github.com/istio/proxy/pull/4628)

- [x] Bump master to version 1.18 in the release-1.18 branches 
  -  BASE_VERSION should update when a new base image is created (vulnerabilities scan fails)
  - UPDATE_BRANCH in update_deps.sh
  - [x] [1.18 istio PR](https://github.com/istio/istio/pull/44615)
  - [x] [1.18 test-infra PR](https://github.com/istio/test-infra/pull/4521) [1.18 PR](https://github.com/istio/test-infra/pull/4677)

- [x] Bump 1.18 to 1.19 in the master branch 
  - Update Istio: 1.18.0 to 1.19.0, 1.18-dev to 1.19-dev, etc: 
  - [x] [1.18 PR](https://github.com/istio/istio/pull/44614)

- [x] Send an announcement on Discuss announcing branch cut complete and PRs can be merged again 

- [x] Ask istio.io team to run job to update to use the 1.18 branch (probably @ericvn) [PR](https://github.com/istio/istio.io/pull/13134)

## Publishing a Release
The next step is to publish a release.  You will need to follow these steps for each beta and release candidate and finally for the final release build.

- [x] Publish a beta release (from https://docs.google.com/document/d/1ilwz-HGSeiRDCFwP5rm-5d_UY6OOYL_nflx7bNggek0/edit#heading=h.qex63c29z2to, builds on release branch are labeled `beta`. Updating below names to reflect that.
  - [x] Run bin/update_deps.sh in istio/istio [PR](https://github.com/istio/istio/pull/44697)
  - [x] Trigger a 1.18 beta.0 build [PR](https://github.com/istio/release-builder/pull/1473)
  - [x] Some verification to verify build is good
    - You can download the build tar from the appropriate directory here: https://gcsweb.istio.io/gcs/istio-prerelease/prerelease/
    - untar the file and install, specifying the image repository. (ex: `./bin/istioctl install --set profile=demo -y --set hub=gcr.io/istio-prerelease-testing`).
    - Run bookinfo from within the untar'd directory
  - [x] Publish a 1.18 beta.0 release

- [x] Send an announcement on Discuss announcing availability of beta.[1.18 Discuss](https://discuss.istio.io/t/announcing-istio-1-18-beta0-branch-cut-completion/15501)

### When to start: 10-14 days before the release date
Use the Publishing a Release instructions above to update_deps, trigger the build, do verification and publish the release

Next, contact Sergii Shapar and ask him to run the long running tests on the build, and any subsequent release candidate builds if there winds up being more than one.

### When to start: 2 weeks ahead of the release date

- [x] Verify the [min](https://github.com/istio/istio/pull/43236) and max K8s version supported for the release.
    - RMs usually determine the range of k8s versions and it appears on the TOC agenda at times. The max version is usually is the current K8s version and the min version is 3 past versions.  

### Generating the release notes
- [x] Prepare the release notes for the release. 

Preparing the release notes for a major release is time-consuming; allow 4-5 hours across several days to factor in time for reviews.  An example PR for the 1.18 release notes is [here](https://github.com/istio/istio.io/pull/13269)

To generate the release notes, first make sure you have up-to-date forks of istio, API and tools repositories checked out to the same folder.  Specifically, they must contain the release-N and release-N-1 branches (where N is the version that you're working on, and N-1 is the previous one), so you must not have only forked the repos' master branches. 

To generate release notes, run the command `./gen-release-notes --notes ../../../istio --oldBranch 1.17.0 --newBranch release-1.18` (against the istio/istio and istio/api repos). Follow the instructions from [here](https://github.com/istio/tools/tree/master/cmd/gen-release-notes).  This will get you text files that contain the contents of what will be the change-notes part of your PR (see the 1.18 PR above for what I mean)

The output of the tool will fail linting.  Once you've copied the content into the right place in your fork of the isio.io repo, run `make lint` then go get a cup of coffee.  Five to ten minutes later, you'll have a list of things that are wrong with the automatically generated content that you'll need to work through.  The biggest source of errors is that there cannot be more than one blank line between any line in the file, headers must be surrounded by blank lines, all istio.io links must be relative and that there will be trailing whitespace that will need removing.  To fix the last of these, run `sed -i 's/[[:space:]]*$//' file.txt`; this will clean up the trailing whitespace that will otherwise drive you slightly mad trying to find (since it's not always quite on the line indicated by the tool)

- [x] Each time you publish a beta/rc/official release build, make sure that the build/helm charts/images are published correctly to [Istio release repo](https://github.com/istio/istio/releases)/[Helm repo](gcr.io/istio-release/charts)/[gcr repo](gcr.io/istio-release/charts).
- [x] After official release, send out an announcement on `announcement` and release slack channel for Istio. Currently Craig Box handles all the twitter announcements related to Istio and he usually sends release related announcements on twitter. Coordinate with him. [discuss](https://discuss.istio.io/t/istio-1-17-is-out/14951), [twitter](https://twitter.com/IstioMesh/status/1625654199336968197?cxt=HHwWioDR8bL4vY8tAAAA)
- [ ] Update the description in the release slack channel to reflect the official release date.

**Comment by pmerrison** [comment]
Automated branching is now largely complete.  There are a few PRs left to merge from step 5, then tomorrow we'll be on to creating the beta images.

**Comment by lisfo4ka** [comment]
hi @pmerriso, could you give any ETA about final release 1.18, please?

**Comment by zirain** [comment]
@lisfo4ka https://github.com/istio/istio/wiki/Istio-Release-1.18

**Comment by lisfo4ka** [comment]
@zirain thanks a lot! We're looking forward for it)
